### PR TITLE
Create setters for substitutions

### PR DESCRIPTION
**Pull request by phanimarupaka**
@pwittrock

### CODE REVIEW

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 1]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	if err != nil {
+		return err
+	}
+
+	// for each ref in values, check if the setter already exists, if not create them
+	for _, value := range c.Values {
+		obj, err := y.Pipe(yaml.Lookup(
+			// get the setter key from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+			// extract io.k8s.cli.setters.image_setter
+			"openAPI", "definitions", strings.Split(value.Ref, "/")[2]))
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: use SplitN so there at at most 3 parts

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 2]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:        strings.Split(value.Ref, ".")[4],
+				Value:       m[value.Marker],
+				SetBy:       "",
+				Description: "Created for substitution " + c.Name,
+				Count:       0,
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: count defaults to 0

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 3]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+			return err
+		}
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:        strings.Split(value.Ref, ".")[4],
+				Value:       m[value.Marker],
+				SetBy:       "",
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: setby defaults to empty

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 1 above)
Comment (pwittrock) [reply kyaml/setters2/settersutil/substitutioncreator.go]: or better yet -- use `TrimPrefix(value.Ref, "#/definitions/")`

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 4]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+
+		if err != nil {
+			return err
+		}
+
+		if obj == nil {
+			sd := setters2.SetterDefinition{
+				// get the setter name from ref. Ex: from #/definitions/io.k8s.cli.setters.image_setter
+				// extract image_setter
+				Name:        strings.Split(value.Ref, ".")[4],
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: Why not use TrimPrefix?

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 5]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+
+// GetStartIndices returns the start indices of all the markers in the pattern
+func (c SubstitutionCreator) GetStartIndices() (map[int]string, error) {
+	inds := make(map[int]string)
+	p := c.Pattern
+	for _, value := range c.Values {
+		m := value.Marker
+		found := false
+		for i, _ := range p {
+			if p[i:int(math.Min(float64(i+len(m)), float64(len(p))))] == m {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: we need to check for overlapping inds and throw an error

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 6]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+				inds[i] = m
+				found = true
+			}
+		}
+		if !found {
+			return nil, errors.New("Unable to find marker " + m + " in the pattern")
+		}
+	}
+	return inds, nil
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: check for overlapping values in inds -- e.g. if there is another map key in the range of i and i + len(inds[i]) then there is an error

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 5 above)
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: can this be simplified to:

```go
if strings.HasPrefix(p[i:], m)
```

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 7]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+		m := value.Marker
+		found := false
+		for i, _ := range p {
+			if p[i:int(math.Min(float64(i+len(m)), float64(len(p))))] == m {
+				inds[i] = m
+				found = true
+			}
+		}
+		if !found {
+			return nil, errors.New("Unable to find marker " + m + " in the pattern")
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: use `kyaml` errors, not `kind` errors -- `errors.Errorf("")`

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 8]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
 		Outputs: []kio.Writer{inout},
 	}.Execute()
 }
+
+// CreateSettersForSubstitution creates the setters for all the references in the substitution
+// values if they don't already exist in openAPIPath file.
+func (c SubstitutionCreator) CreateSettersForSubstitution(openAPIPath string) error {
+	b, err := ioutil.ReadFile(openAPIPath)
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: consider using https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/types.go#L123

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 9]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+		}
+	}
+	return nil
+}
+
+// GetValuesForMarkers parses the pattern and field value to derive values for the
+// markers in the pattern string. Returns error if the marker values can't be derived
+func (c SubstitutionCreator) GetValuesForMarkers() (map[string]string, error) {
+	m := make(map[string]string)
+	inds, _ := c.GetStartIndices()
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: is ignoring the error the right thing to do here?

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 10]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	s:=c.FieldValue
+	p:=c.Pattern
+	i:=0
+	j:=0
+	for i< len(s) && j< len(p) {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: can you describe the algorithm in a comment?

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 11]
```diff
@@ -62,3 +73,106 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+			}
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.New("Same marker is found to have different values in field value.")
+			}
+			m[marker] = value
+			j+=len(marker)
+		} else {
+			if s[i] != p[j] {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: move this check to the top of the loop and break early for less complexity --

```go
if _, found := inds[j]; !found && s[i] != p[i] {
  return nil, errors.
}
```

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 6 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Is this the case where one marker is substring of other? The algorithm handles it but throws generic error. There will be many such cases and question is to what level we should report specific errors. Algorithm is more stringent and looks for perfect match of pattern and field value. Can you quote an example and I can verify.

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 12]
```diff
@@ -62,3 +73,114 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	inds, err := c.GetStartIndices()
+	if err != nil {
+		return nil, err
+	}
+	s:=c.FieldValue
+	p:=c.Pattern
+	i:=0
+	j:=0
+	// iterate s, p with indices i, j respectively and when j hits the index of a marker,
+	// freeze j and iterate i till we find the character just after marker and capture the string
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: don't we need to match the substring after the marker, not just the character?  what if the character after the marker is part of the value -- e.g. pattern `/projects/PROJECT_SETTER/zones/ZONE_SETTER` with value `/projects/project1//zones/zone1` -- this would stop the capture when it its the first `/`, but should stop when it hits the second

Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Done

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 4 above)
Comment (pwittrock) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Looks like it still isn't using TrimPrefix?

Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Sorry. Missed it. Done now.

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 6 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Please refer to unit test TestGetValuesForMarkersSubStngMarkers

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: I think this method is more explicit is what I feel.

File: kyaml/go.sum [hunk 13]
```diff
@@ -1,56 +1,163 @@
...
 gopkg.in/yaml.v2 v2.2.2/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
 gopkg.in/yaml.v2 v2.2.4 h1:/eiJrUcujPVeJ3xlSWaiNi3uSVmDGBK1pDHUHAnao1I=
 gopkg.in/yaml.v2 v2.2.4/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
-gopkg.in/yaml.v3 v3.0.0-20191026110619-0b21df46bc1d h1:LCPbGQ34PMrwad11aMZ+dbz5SAsq/0ySjRwQ8I9Qwd8=
-gopkg.in/yaml.v3 v3.0.0-20191026110619-0b21df46bc1d/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
+gopkg.in/yaml.v2 v2.2.7 h1:VUgggvou5XRW9mHwD/yXxIYSMtY0zoKQf/v226p2nyo=
+gopkg.in/yaml.v2 v2.2.7/go.mod h1:hI93XBmqTisBFMUTm0b8Fm+jr3Dg1NNxqwp+5A1VGuI=
+gopkg.in/yaml.v3 v3.0.0-20191120175047-4206685974f2 h1:XZx7nhd5GMaZpmDaEHFVafUZC7ya0fuo7cSJ3UCKYmM=
+gopkg.in/yaml.v3 v3.0.0-20191120175047-4206685974f2/go.mod h1:K4uyk7z7BCEPqu6E+C64Yfv1cQ7kz7rIZviUmN+EgEM=
+k8s.io/apimachinery v0.17.0 h1:xRBnuie9rXcPxUkDizUsGvPf1cnlZCFu210op7J7LJo=
```
Comment (pwittrock) [inline kyaml/go.sum]: we can't introduce a dependency on libraries from kubernetes/kubernetes

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 8 above)
Comment (pwittrock) [reply kyaml/setters2/settersutil/substitutioncreator.go]: At least use this: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/types.go#L104

No need to copy-paste these lines around.

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 14]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+		return nil, err
+	}
+	s := c.FieldValue
+	p := c.Pattern
+	i := 0
+	j := 0
+	// iterate s, p with indices i, j respectively and when j hits the index of a marker, freeze j and iterate
+	// i and capture string till we find the substring just after current marker and before next marker
+
+	// Ex: s = "something/ubuntu:0.1.0", p = "something/IMAGE::VERSION", till j reaches 10
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: the value doesn't have `::`?

Comment (pwittrock) [reply kyaml/setters2/settersutil/substitutioncreator.go]: probably should pick a reasonable value -- e.g.  just `:`

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 15]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	// just proceed i and j and check if s[i]==p[j]
+	// when j is 10, freeze j and move i till it sees substring '::' which derives IMAGE = ubuntu and so on.
+	for i < len(s) && j < len(p) {
+		if marker, ok := indices[j]; ok {
+			value := ""
+			e := j + len(marker)
+
+			for i < len(s) && (e == len(p) ||
+				s[i:min(len(s), i+lenToNextMarker(indices, e))] != p[e:min(e+lenToNextMarker(indices, e), len(p))]) {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: can you pull these out into variables so the reader understands what they represent?

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 16]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+				value += string(s[i])
+				i++
+			}
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.Errorf("Same marker is found to have different values in field value.")
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: don't capitalize errors or have punctuation: https://github.com/golang/go/wiki/Errors

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 17]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+			}
+			m[marker] = value
+			j += len(marker)
+		} else {
+			if s[i] != p[j] {
+				return nil, errors.Errorf("Unable to derive values for markers. Create setters for all markers and then try again.")
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: same comment about error style

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 18]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+				value += string(s[i])
+				i++
+			}
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.Errorf("Same marker is found to have different values in field value.")
+			}
+			m[marker] = value
+			j += len(marker)
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: its pretty hard to follow these lines -- at a minimum these should have comments explain why this value is being incremented

WDYT of factoring whole thing into a struct with functions?

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 19]
```diff
@@ -62,3 +74,138 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+
+// GetValuesForMarkers parses the pattern and field value to derive values for the
+// markers in the pattern string. Returns error if the marker values can't be derived
+func (c SubstitutionCreator) GetValuesForMarkers() (map[string]string, error) {
+	m := make(map[string]string)
+	indices, err := c.GetStartIndices()
+	if err != nil {
+		return nil, err
+	}
+	s := c.FieldValue
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: are there more expressive variable names we could use? -- e.g. `i` is `valueIndex` and `j` is `patternIndex` or something

**Comment by pwittrock** [comment]
plz check travis failures:

:100644 100644 0591d04c2b9229e6e97dcb3927c02df5e52bed60 368198f7ef80eb8758e5530fdcebe671231c3bf1 M	cmd/config/go.mod
:100644 100644 00d5c229a61a2f5a76a8b45882a1ac31728ebe16 7e73dac94f6173733eca3830c31c4074b209a69c M	kyaml/go.mod
The command "./travis/check-go-mod.sh" exited with 1.

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 19 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Will do.

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 15 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Will do.

File: kyaml/setters2/settersutil/substitutioncreator.go (hunk 8 above)
Comment (phanimarupaka) [reply kyaml/setters2/settersutil/substitutioncreator.go]: Ahh thanks for pointing.

File: kyaml/go.sum (hunk 13 above)
Comment (phanimarupaka) [reply kyaml/go.sum]: I push only changes added by go mod tidy

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 20]
```diff
@@ -62,3 +73,144 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+func (c SubstitutionCreator) extractValueForMarker(fvInd *int, fv string, patternInd int, indices map[int]string) string {
+	value := ""
+	lm := lenTillNextMarker(indices, patternInd)
+
+	// keep appending chars to value till we find the substring between 2 markers in fv
+	// In example fv = "something/ubuntu:0.1.0", pattern = "something/IMAGE:VERSION",
+	// proceed till we find substring ":" in fv which makes value = ubuntu for marker IMAGE
+	for *fvInd < len(fv) && (patternInd == len(c.Pattern) ||
+		fv[*fvInd:min(len(fv), *fvInd+lm)] != c.Pattern[patternInd:min(patternInd+lm, len(c.Pattern))]) {
+		value += string(fv[*fvInd])
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: // add the current character to the extracted value

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 21]
```diff
@@ -62,3 +73,144 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+		*fvInd++
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: // increment the field value index so we look at the next character

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 22]
```diff
@@ -62,3 +73,144 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	for fvInd < len(fv) && patternInd < len(pattern) {
+		// if we hit marker index, extract its corresponding value
+		if marker, ok := indices[patternInd]; ok {
+			// increment the patternInd to end of marker. This helps us to extract the substring before next marker.
+			patternInd += len(marker)
+			value := c.extractValueForMarker(&fvInd, fv, patternInd, indices)
+			// if marker is repeated in the pattern, make sure that the corresponding values
+			// are same and throw error if not.
+			if prevValue, ok := m[marker]; ok && prevValue != value {
+				return nil, errors.Errorf("same marker is found to have different values in field value")
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: say the 2 values

File: kyaml/setters2/settersutil/substitutioncreator_test.go [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+	if !assert.Error(t, err) {
+		t.FailNow()
+	}
+
+	if !assert.Equal(t, err.Error(), "unable to derive values for markers, create setters for all markers and then try again") {
+		t.FailNow()
+	}
+}
+
+func TestGetValuesForMarkersSubStngMarkers(t *testing.T) {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator_test.go]: add a test for

	c := SubstitutionCreator{
		Pattern:    "something/AA:something/A,
		Values:    "markers are A and AA"
                 FieldValue: "something/fofo:something/ba,
	}

	c := SubstitutionCreator{
		Pattern:    "something/A:something/AA,
		Values:    "markers are A and AA"
                 FieldValue: "something/fofo:something/ba,
	}

also

non-separated: MARKER_1MARKER_2

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 24]
```diff
@@ -62,3 +73,171 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+
+	// Ex: fv = "something/ubuntu:0.1.0", pattern = "something/IMAGE:VERSION", till patternInd reaches 10
+	// just proceed fvInd and patternInd and check if fv[fvInd]==pattern[patternInd] when patternInd is 10,
+	// freeze patternInd and move fvInd till it sees substring ':' which derives IMAGE = ubuntu and so on.
+	for fvInd < len(fv) && patternInd < len(pattern) {
+		// if we hit marker index, extract its corresponding value
+		if marker, ok := indices[patternInd]; ok {
+			// increment the patternInd to end of marker. This helps us to extract the substring before next marker.
+			patternInd += len(marker)
+			value, fvIndIncremented := c.extractValueForMarker(fvInd, fv, patternInd, indices)
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: consider using `var value string`

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 25]
```diff
@@ -62,3 +73,171 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	return indices, nil
+}
+
+// validateMarkerPositions takes the indices map and checks if any of 2 markers not have delimiters
+// and returns error
+func validateMarkerPositions(indices map[int]string) error {
+	for k1, v1 := range indices {
+		for k2, v2 := range indices {
+			if k1 != k2 && k1+len(v1) == k2 {
+				return errors.Errorf("markers %s and %s are found to have no delimiters between them", v1, v2)
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: they must be manually created

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 26]
```diff
@@ -62,3 +73,171 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+	return fv[valueStartInd:fvInd], fvInd
+}
+
+// substrOfLen takes a string, start index and length and returns substring of given length
+// or till end of string
+func substrOfLen(str string, startInd int, length int) string {
+	return str[startInd:min(len(str), startInd+length)]
+}
+
+// lenToNextMarker takes in the indices map, an index and returns the distance to
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: update

File: kyaml/setters2/settersutil/substitutioncreator.go [hunk 27]
```diff
@@ -62,3 +73,171 @@ func (c SubstitutionCreator) Create(openAPIPath, resourcesPath string) error {
...
+// next marker start index in indices map
+func strTillNextMarker(m map[int]string, j int, pattern string) string {
+	res := math.MaxInt32
+	for k := range m {
```
Comment (pwittrock) [inline kyaml/setters2/settersutil/substitutioncreator.go]: // find the index of the next marker (e.g. the smallest value in map greater than j)
